pkg/mirror/cached_upstream.go

```diff
@@ -0,0 +1,172 @@
+		provider.Hostname, provider.Namespace, provider.Name, provider.Version)
+}
+
+// Estimates the size in bytes of an object via JSON marshaling (usefull for cache Weigther func)
+func estimateSize(data interface{}) int {
+	bytes, err := json.Marshal(data)
+	if err != nil {
+		return 0
+	}
+	return len(bytes)
+}
+
+// Implements upstreamProvider's listProviderVersions method, with caching
+func (c *cachedUpstreamProvider) listProviderVersions(ctx context.Context, provider *core.Provider) (*core.ProviderVersions, error) {
+	key := buildVersionsKey(provider)
+
+	// Try to get from cache
+	if entry, ok := c.cache.GetIfPresent(key); ok {
+		if versions, ok := entry.data.(*core.ProviderVersions); ok {
+			return versions, nil
+		}
+	}
+
+	// Cache miss - call upstream
+	versions, err := c.upstream.listProviderVersions(ctx, provider)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store in cache
+	entry := &cacheEntry{
+		data:      versions,
+		timestamp: time.Now(),
+		sizeBytes: estimateSize(versions),
+	}
+	c.cache.Set(key, entry)
+
+	return versions, nil
+}
+
+// Implements upstreamProvider's getProvider method, with caching
+func (c *cachedUpstreamProvider) getProvider(ctx context.Context, provider *core.Provider) (*core.Provider, error) {
+	key := buildProviderKey(provider)
+
+	// Try to get from cache
+	if entry, ok := c.cache.GetIfPresent(key); ok {
+		if prov, ok := entry.data.(*core.Provider); ok {
+			return prov, nil
+		}
+	}
+
+	// Cache miss - call upstream
+	prov, err := c.upstream.getProvider(ctx, provider)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store in cache
+	entry := &cacheEntry{
+		data:      prov,
+		timestamp: time.Now(),
+		sizeBytes: estimateSize(prov),
+	}
+	c.cache.Set(key, entry)
+
+	return prov, nil
+}
+
+// Implements upstreamProvider's shaSums method, with caching
+func (c *cachedUpstreamProvider) shaSums(ctx context.Context, provider *core.Provider) (*core.Sha256Sums, error) {
+	key := buildShaSumsKey(provider)
+
+	// Try to get from cache
+	if entry, ok := c.cache.GetIfPresent(key); ok {
+		if sums, ok := entry.data.(*core.Sha256Sums); ok {
+			return sums, nil
+		}
+	}
+
+	// Cache miss - call upstream
+	sums, err := c.upstream.shaSums(ctx, provider)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store in cache
+	entry := &cacheEntry{
+		data:      sums,
+		timestamp: time.Now(),
+		sizeBytes: estimateSize(sums),
+	}
+	c.cache.Set(key, entry)
+
+	return sums, nil
+}
+
+// Creates a new upstream provider wrapper with caching
+func newCachedUpstreamProvider(upstream upstreamProvider, config CacheConfig) (*cachedUpstreamProvider, error) {
+	// Convert MB to bytes
+	maxWeightBytes := config.MaxSizeMB * 1024 * 1024
+
+	// Configure otter cache
+	opts := &otter.Options[string, *cacheEntry]{
+		MaximumWeight: maxWeightBytes,
+		Weigher: func(key string, value *cacheEntry) uint32 {
+			// Weight = size of key + size of value
+			return uint32(len(key) + value.sizeBytes)
+		},
```

Comment:
Shouldn't we be using `MaximumSize` ([docs](https://maypok86.github.io/otter/user-guide/v2/features/eviction/#size-based)) instead?